[dynamo] fix functools.wraps on nested functions

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #114279

Updated version of #108885 addressing the review. In this PR:
- We add a VT.can_reconstruct utility that checks if VT.reconstruct()
  does something.
- If functools.wraps(fn) is passed a `fn` that either has a source or
  has .can_reconstruct() == True, then we stash the source (or the VT)
- Later on, we use the source (or VT.reconstruct) to actually
  reconstruct the object in codegen.

Test Plan:
- New tests

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

zou3519 committed:
[dynamo] fix functools.wraps on nested functions

Updated version of #108885 addressing the review. In this PR:
- We add a VT.can_reconstruct utility that checks if VT.reconstruct()
  does something.
- If functools.wraps(fn) is passed a `fn` that either has a source or
  has .can_reconstruct() == True, then we stash the source (or the VT)
- Later on, we use the source (or VT.reconstruct) to actually
  reconstruct the object in codegen.

Test Plan:
- New tests

[ghstack-poisoned]

jon-chuang commented on torch/_dynamo/variables/misc.py:
```diff
@@ -852,12 +852,16 @@ def call_function(
             and not kwargs
             and len(args) == 1
-            and args[0].source
+            and (args[0].source is not None or args[0].can_reconstruct(tx.output.root_tx))
```
Just curious, does it need to be the root_tx?

jon-chuang replied:
I think just tx is fine right?

zou3519 replied:
I get some assertions when it is not the root tx. The moral reason we're using the root tx is because one only reconstructs VT at the root tx (instead of InliningInstructionTranslators)

jon-chuang commented:
I think this makes sense. We should not fail to wrap a function just because it has no source.

Is there a plan to extend this further to handle all/most `NestedUserFunctionVariable`s?

zou3519 committed:
Update on "[dynamo] fix functools.wraps on nested functions"


Updated version of #108885 addressing the review. In this PR:
- We add a VT.can_reconstruct utility that checks if VT.reconstruct()
  does something.
- If functools.wraps(fn) is passed a `fn` that either has a source or
  has .can_reconstruct() == True, then we stash the source (or the VT)
- Later on, we use the source (or VT.reconstruct) to actually
  reconstruct the object in codegen.

Test Plan:
- New tests

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

zou3519 commented:
> I think this makes sense. We should not fail to wrap a function just because it has no source.
> 
> Is there a plan to extend this further to handle all/most `NestedUserFunctionVariable`s?
> 
> It can be follow up work.

No plans from me, I am just trying to get as much working so that we are able to trace through torch.vmap internals.

Skylion007 commented on torch/_dynamo/variables/functions.py:
```diff
@@ -381,7 +381,10 @@ def __init__(
         self.closure_scope = closure_scope
-        self.wraps_source = wraps_source
+        # Either a source or a VT with .can_reconstruct() == True
+        self.wrapped_reconstructible: Optional[
```
Can this type hint be moved to the `__init__` definition, or does it need to be here as well?

zou3519 committed:
Update on "[dynamo] fix functools.wraps on nested functions"


Updated version of #108885 addressing the review. In this PR:
- We add a VT.can_reconstruct utility that checks if VT.reconstruct()
  does something.
- If functools.wraps(fn) is passed a `fn` that either has a source or
  has .can_reconstruct() == True, then we stash the source (or the VT)
- Later on, we use the source (or VT.reconstruct) to actually
  reconstruct the object in codegen.

Test Plan:
- New tests

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

zou3519 committed:
Update on "[dynamo] fix functools.wraps on nested functions"


Updated version of #108885 addressing the review. In this PR:
- We add a VT.can_reconstruct utility that checks if VT.reconstruct()
  does something.
- If functools.wraps(fn) is passed a `fn` that either has a source or
  has .can_reconstruct() == True, then we stash the source (or the VT)
- Later on, we use the source (or VT.reconstruct) to actually
  reconstruct the object in codegen.

Test Plan:
- New tests

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

zou3519 commented:
@pytorchbot merge
